### DISCUSSION TITLE
Program change issue

### DISCUSSION THREAD

**Discussion by sebastiandior** [RomanKubiak/ctrlr#139] [General]
Hi, 
I am VERY very novice at programming.
I am trying to create a small easy programchange button set, up and down. 
When I set the value to 1 or -1 it only works once, IE if i click the button that as the program change set, I will get it change the program one value higher. Then if i click it again, it will send the value negative and go back to where I was. 
I am pretty sure I am not doing this correctly, but I would love some help on how to get this done, as I have some external rack gear  that I would like to send programchange to. 
Thanks

**Comment by sebastiandior** [comment]
The idea comes from the Midi Mouse device that sends program change only. If that helps.

**Reply by Ted-juh** [reply]
I'm at work right now so I can't write a proper code on my phone. Too cumbersome. But you need some lua script to do what you want to do as you might have guessed. 

Basically it comes down to this: You need two buttons and a label. The label reads it's content from a lua-table. The two buttons add or extract a number to go through that table to read the content or midi message to send.

Another option could be a combobox or a listbox. Easier to handle but the listbox might take up more space in your panel.

**Comment by dobo365** [comment]
Hi and welcome to the Ctrlr world Sebastian!
Did you have a look in the old forum? I'm pretty sure this has been explained several times.
Can you mention the synth you are using?

Even as experienced Ctrlr user/developer, I have never tried the embedded Program change message.
Will try on my Pro-2 this afternoon.

Otherwise, I can give you some code as @Tedjuh1976 mentioned (or he will provide you some code ;-) )

Program and Bank change messages are standard midi messages on all synths.
You can search for that on internet.
But then indeed, you will have to put that into some Lua code associated to your buttons.

**Reply by sebastiandior** [reply]
I doubt if i'm up to the task of writing lua.

**Comment by Ted-juh** [comment]
A small panel that has a lot of code for someone who doesn't know Lua. There are three scripts: Init, buttonUp, buttonDown. The init file is to set some values and modulators when the panel is loaded. The buttonUp/ buttonDown scripts are to handle the math. It can also send the midi message but you need to comment that out. For your convenience, there is also a midi channel selector. 

One thing I'm not very keen on is that the message gets sent every time you click an up/ down button. This makes the device change programs all the way until you reached the desired number. So, a lot of unnecessary requests if you ask me. Better would be to work with another combo to go straight to the number you want. Or a separate button that sends the message apart from the selection buttons. But hey, that's up to you. 

[programChange1.zip](https://github.com/RomanKubiak/ctrlr/files/5862025/programChange1.zip)

**Reply by dobo365** [reply]
@sebastiandior now you are on the path to learn Lua and Ctrlr further! 
Start with the code provided by @Tedjuh1976 then explore gradually further.
@Tedjuh1976 thx man!

**Reply by sebastiandior** [reply]
Thank you, i will study this.

**Reply by sebastiandior** [reply]
So the same thing happens with this, as it was not set to programchange when i opened it. But I set it to program change and it only works once.  [https://imgur.com/a/RQx0iDs]  (https://imgur.com/a/RQx0iDs)

**Comment by Ted-juh** [comment]
Open the "original" panel I uploaded a few messages earlier. Don't.. I said "DON'T" touch that "program change" property. :) Instead, go to "Panel" in the menubar, go to "Lua Editor".  It opens a new window, that is where you can see the code I wrote. Now, I dare you to press the up-button a few times, throw a down-button in the mix and you'll see a message in the bottom of the "Lua Editor" window that says something like "LUA>> message to send = C0 12". That, my friend, is the program change that gets generated if you press up or down. 

The panel doesn't send the message because it is out-commented in the code. When opening the "lua editor" it shows you the scripts on the left side of the editor. Double-click "buttonUp" if it isn't showing already. Look to the right of the "Lua Editor", that is where you see the code itself. You will see a line of code (line 30?) that says:

--		panel:sendMidiMessageNow(CtrlrMidiMessage(msg))

You need to delete the -- in the beginning of that line to let the panel send the program change. You have to do the same in the buttonDown script. Try it and report if it works now. Don't forget to set the midi in/ out channels in the settings.

**Reply by sebastiandior** [reply]
Ok wonderful, it works, but it is offset by one, when I says 30, It is showing 29 on my device. Is there a way to offset that? Is there a button to that can be created to toggle that offset? I hope that makes sense.

**Reply by Ted-juh** [reply]
Sorry, there is a mixup in my code. I added an +1 value that was unnecessary. So the "newValue" is not needed. 

I see Mr. Dnaldoog had a panel without code. Big thumbs up. I didn't even know that it was possible without Lua. So, as you can see, there are multiple roads that lead to Rome.

**Comment by dnaldoog** [comment]
@sebastiandior Here's a panel that changes the program without lua.
![changeProgram139](https://user-images.githubusercontent.com/76831683/105675513-95b81100-5f2c-11eb-9310-a31ed8581e50.png)

Regards,

[Change Program_139_0_GitHub_2021-01-25_16-39.zip](https://github.com/RomanKubiak/ctrlr/files/5864949/Change.Program_139_0_GitHub_2021-01-25_16-39.zip)

**Reply by sebastiandior** [reply]
so it looks like you did the work in the component section of the panel? Very cool, it works well.
